### PR TITLE
[webui] Refactor project save_meta action

### PR DESCRIPTION
**Pull request by ChrisBr**
- Removed ActiveXML call from save_meta action and using Project model
  functions instead
- Moved functions from controller to model
- Handle permissions in policy

### CODE REVIEW

**Comment by ChrisBr** [comment]
@bgeuken @mdeniz @adrianschroeter please review!

I'll add tests if you agree with my approach!

File: src/api/app/controllers/source_controller.rb [hunk 1]
```diff
@@ -167,27 +167,26 @@ def render_project_verboseproductlist
...
       render_error :status => 403, :errorcode => 'delete_project_no_permission',
                    :message => "Permission denied (delete project #{project_name})"
       return
     end
-    pro.can_be_deleted?
-
-    # find linking repos
-    private_check_and_remove_repositories(params, pro.repositories) or return
+    project.can_be_deleted?
+    full_remove = (params[:remove_linking_repositories] && !params[:remove_linking_repositories].empty?)
```
Comment (bgeuken) [inline src/api/app/controllers/source_controller.rb]: !params[:remove_linking_repositories][.blank?](http://apidock.com/rails/Object/blank%3F);-)

File: src/api/app/controllers/source_controller.rb [hunk 2]
```diff
@@ -475,141 +474,98 @@ class ProjectReadAccessFailure < APIException
...
     begin
-      prj = Project.get_by_name rdata['name']
+      project = Project.get_by_name(request_data['name'])
     rescue Project::UnknownObjectError
-      prj = nil
+      project = nil
     end
 
     # projects using remote resources must be edited by the admin
     unless User.current.is_admin?
```
Comment (bgeuken) [inline src/api/app/controllers/source_controller.rb]: I would move this block up to line 486. Then we have 2 logical groups: parsing the xml and processing it and fetching the project and processing that.

File: src/api/app/controllers/source_controller.rb [hunk 3]
```diff
@@ -475,141 +474,98 @@ class ProjectReadAccessFailure < APIException
...
+
+    error = Project.validate_repository_xml_attribute(request_data, project_name)
+    if error[:error]
+      raise RepositoryAccessFailure, error[:error]
+    end
+
+    if project
+      new_repo_names = project.get_repositories_from_xml(request_data)
+      remove_repositories = project.get_removed_repositories(new_repo_names)
+      full_remove = (params[:remove_linking_repositories] && !params[:remove_linking_repositories].empty?)
```
Comment (bgeuken) [inline src/api/app/controllers/source_controller.rb]: .blank? ;-)

File: src/api/app/controllers/source_controller.rb [hunk 4]
```diff
@@ -475,141 +474,98 @@ class ProjectReadAccessFailure < APIException
...
-      if tprj.class == Project and tprj.disabled_for?('access', nil, nil) and
-          !FlagHelper.xml_disabled_for?(rdata, 'access')
-        raise ProjectReadAccessFailure.new "project links work only when both projects have same read access " +
-                                           "protection level: #{project_name} -> #{tproject_name}"
+  def check_and_remove_repositories!(repositories, full_remove, force = false)
+    error = Project.check_repositories(repositories)
+    if error[:error] && !force
+      raise RepoDependency, error[:error]
+    else
+      # TODO check wie das vorher war
```
Comment (bgeuken) [inline src/api/app/controllers/source_controller.rb]: Looks like a leftover

File: src/api/app/controllers/source_controller.rb [hunk 5]
```diff
@@ -475,141 +474,98 @@ class ProjectReadAccessFailure < APIException
...
-      tprj = Project.get_by_name(tproject_name)
 
-      # The read access protection for own and linked project must be the same.
-      # ignore this for remote targets
-      if tprj.class == Project and tprj.disabled_for?('access', nil, nil) and
-          !FlagHelper.xml_disabled_for?(rdata, 'access')
-        raise ProjectReadAccessFailure.new "project links work only when both projects have same read access " +
-                                           "protection level: #{project_name} -> #{tproject_name}"
+  def check_and_remove_repositories!(repositories, full_remove, force = false)
+    error = Project.check_repositories(repositories)
```
Comment (bgeuken) [inline src/api/app/controllers/source_controller.rb]: With ```error = Project.check_repositories(repositories) unless force``` we can spare unnecessary check repository calls

File: src/api/app/controllers/source_controller.rb [hunk 6]
```diff
@@ -475,141 +474,98 @@ class ProjectReadAccessFailure < APIException
...
+    if error[:error] && !force
+      raise RepoDependency, error[:error]
+    else
+      # TODO check wie das vorher war
+      error = Project.remove_repositories(repositories, full_remove)
+      if error[:error] && !force
```
Comment (bgeuken) [inline src/api/app/controllers/source_controller.rb]: same here

File: src/api/app/controllers/webui/project_controller.rb [hunk 7]
```diff
@@ -666,14 +668,32 @@ def meta
...
-      render text: e.summary, :status => 400, :content_type => 'text/plain'
-      return
+    request_data = Xmlhash.parse(params[:meta])
+
+    errors = []
+    errors << Project.validate_link_xml_attribute(request_data, @project.name)[:error]
+    errors << Project.validate_maintenance_xml_attribute(request_data)[:error]
+    errors << Project.validate_repository_xml_attribute(request_data, @project.name)[:error]
+
+    full_remove = params[:remove_linking_repositories] && !params[:remove_linking_repositories].empty?
```
Comment (bgeuken) [inline src/api/app/controllers/webui/project_controller.rb]: blank?

File: src/api/app/controllers/source_controller.rb (hunk 6 above)
Comment (bgeuken) [inline src/api/app/controllers/source_controller.rb]: Alright, this really removes repos. Nevermind then

**Comment by bgeuken** [comment]
LGTM, beside the nitpicks. Though someone else should double check

File: src/api/app/controllers/source_controller.rb [hunk 8]
```diff
@@ -475,141 +474,98 @@ class ProjectReadAccessFailure < APIException
...
-    path = request.path_info
-    path += build_query_from_hash(params, [:user, :comment, :rev])
-    #allowed = false
     request_data = request.raw_post
 
     # permission check
-    rdata = Xmlhash.parse(request_data)
-    if rdata['name'] != project_name
-      raise ProjectNameMismatch.new "project name in xml data ('#{rdata['name']}) does not match resource path component ('#{project_name}')"
+    request_data = Xmlhash.parse(request_data)
```
Comment (mdeniz) [inline src/api/app/controllers/source_controller.rb]: I  would change it to  
```
request_data = Xmlhash.parse(request.raw_post)
```
and then you can delete the line number 479 (```request_data = request.raw_post```)

**Comment by mdeniz** [comment]
LGTM, beside the nitpicks too.

**Comment by ChrisBr** [comment]
@mdeniz @bgeuken fixed your suggestions!

@adrianschroeter please also have a brief look at this one!

**Comment by ChrisBr** [comment]
``
[Coveralls] Submitting to https://coveralls.io/api/v1
[Coveralls] Job #4420.2
[Coveralls] https://coveralls.io/jobs/8284456
Coverage is at 83.82%.
Coverage report sent to Coveralls
``

Travis submitted coverage to coveralls, no idea why the status is still pending in this PR...

@bgeuken @mdeniz added tests and fixed your suggestions, please review!

**Comment by mdeniz** [comment]
LGTM. Coveralls seems to be not working.. my PRs are in the same status.

**Comment by bgeuken** [comment]
LGTM

**Pull request merged**